### ISSUE TITLE
AtlasTexture Region and Margin ignored for Line2D

### ISSUE THREAD

**Issue by nazjun** [godotengine/godot#43268]
**Godot version:**
v3.2.3.stable.mono.official

**OS/device including version:**
Microsoft Windows 10 Pro, version 10.0.18363 Build 18363

**Issue description:**
It's all in the title: Line2D nodes don't make use of AtlasTexture Region and Margin properties and instead use the full texture.
This makes using them with sprite sheets a hassle as you can't specify what rect is visible.
[It seems that this issue has existed for at least a couple years now, judging from this Q&A page.](https://godotengine.org/qa/36054/how-to-animate-a-line2d)

**Steps to reproduce:**
1. Create a Line2D node in the editor and add 2 or more points
2. Set its Texture to an AtlasTexture
3. Modify the Region and/or Margin properties; note how nothing changes on the line

**Minimal reproduction project:**
[AtlasBug.zip](https://github.com/godotengine/godot/files/5471781/AtlasBug.zip)

**Comment by Zylann** [comment]
This issue has the same cause as the following areas of the engine:

- Polygon2D (https://github.com/godotengine/godot/issues/25049)
- MeshInstance and MultiMeshInstance (https://github.com/godotengine/godot-proposals/issues/1708)
- Shaders (https://github.com/godotengine/godot/issues/25895)
- NinePatchRect (under certain configurations)

And the reason is simple: `AtlasTexture` is NOT a texture. It inherits `Texture` but that's only a facade. It barely just holds a reference to a real texture, with a rectangle, and just so happen to work if used on *some* simple 2D nodes because these nodes *happen* to generate their simple quads with special UVs.
For anything more complex, it would be a shitshow of geometry workarounds to implement throughout the engine. Basically, it was never made to work together with other types of nodes. It's meant only for nodes that render simple *quads*.

I believe if we want to solve this, either we have to:
- Make it clear that this kind of "texture" can only be used with certain 2D nodes, and that it's not a bug, but an expected limitation
- Or make it so this kind of texture is handled *at rendering engine level* in such a way the `sampler2D` would be bound to an image region (instead of the whole image), which would make things work naturally without any UV hacks, even with repeat or clamp flags on (does Vulkan allow to do this now?).

**Comment by bsil78** [comment]
@Zylann AtlasTexture is indeed a Texture in users point of view ; how it is implemented should match this expectation. In fact, the only need of the whole source image is only for region capture, after that capture the whole awaited behavior of AtlasTexture is that one of a regular basic texture. It is now clearly a mismatching implementation problem, as this hard behavior link is now commonly recognized by users as true and ancored in usages exoectations.

**Comment by OS-of-S** [comment]
> [@Zylann](https://github.com/Zylann) how it is implemented should match this expectation

Agree. Imagine to make a whole game but have no opportunity to optimize it's textures becouse atlases break everything.

**Comment by FoxyFox909** [comment]
> [@Zylann](https://github.com/Zylann) AtlasTexture is indeed a Texture in users point of view ; how it is implemented should match this expectation. In fact, the only need of the whole source image is only for region capture, after that capture the whole awaited behavior of AtlasTexture is that one of a regular basic texture. It is now clearly a mismatching implementation problem, as this hard behavior link is now commonly recognized by users as true and ancored in usages expectations.

Can confirm. Am user who just tried using an AtlasTexture for a Line2D. After reading here, I understand the limitation and it seems reasonable, but my initial user instinct was to expect it to "Just work". The class description should probably be expanded to mention explicitly that it simply does not work with nodes like Line2D (explicitly mentioning Line2D), tiling aside since LineTextureMode set to stretch mode does not work as one would expect, either.

**Comment by Anime37** [comment]
Just commenting because i got confused with this myself. For now I guess will have to make a separate image file.
